### PR TITLE
[Feature] (batch write part1) Implement the basic workflow on FE side

### LINKED ISSUE #51085: Support batching of stream loads on the server side
- [x] https://github.com/StarRocks/starrocks/pull/52910
- [x] https://github.com/StarRocks/starrocks/pull/52556
- [x] https://github.com/StarRocks/starrocks/pull/52490
- [x] https://github.com/StarRocks/starrocks/pull/52124

## Feature request

**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by banmoy**
## Why I'm doing:
The first PR to support batching stream load on the server side. This pr mainly implements the basic workflow on the FE side. 

## What I'm doing:
The core logic
* http client sends http request to FE, and FE will redirect it a coordinator BE. This process is similar to the normal stream load, but there is difference for how to choose the BE. stream load will carry headers `enable_batch_write` to enable batching on the server side
```
curl --location-trusted -u <username>:<password> \
    -H "Expect:100-continue" \
    -H "format:json" \
    -H "enable_batch_write:true" \
    -H "batch_write_async:true" \
    -H "batch_write_interval_ms:1000" \
    -H "batch_write_parallel:4" \
    -T example1.json -XPUT \
    http://<fe_host>:<fe_http_port>/api/mydatabase/table1/_stream_load
```
* coordinator BE sends thrift request to FE to execute a load. This process is also similar to the normal stream load, but there is difference for how to execute the load plan
* FE will maintain a set of coordinator backends for isomorphic stream loads which is implemented by `CoordinatorBackendAssigner`. `isomorphic` means those stream loads have the same parameters. FE will randomly redirect each stream load to one of those BE so that these stream loads can be batched. The number of backends is controlled by parameter `batch_write_parallel`
* For the load plan, it's almost the same as that of [parallel transaction stream load](https://github.com/StarRocks/starrocks/pull/12182). It uses pipeline engine and can have multiple parallelisms. The main change is that we need deploy it to the coordinator backends assigned by `CoordinatorBackendAssigner`

There will be other PRs to implement the following functions
* complete the logic of coordinator assigner
* the logic on the BE side
* enhance observability

Fixes [#51085](https://github.com/StarRocks/starrocks/issues/51085)

## What type of PR is this:

- [ ] BugFix
- [X] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Review by tracymacding — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/load/batchwrite/LoadExecutor.java [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        if (db == null) {
+            throw new LoadException(String.format("Database %s does not exist", tableId.getDbName()));
+        }
+
+        Table table;
+        Locker locker = new Locker();
+        locker.lockDatabase(db.getId(), LockType.READ);
+        try {
+            table = GlobalStateMgr.getCurrentState()
+                    .getLocalMetastore().getTable(db.getFullName(), tableId.getTableName());
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/load/batchwrite/LoadExecutor.java:235]: better to use `db.getTable()` and lock is not necessary.

File: fe/fe-core/src/main/java/com/starrocks/load/batchwrite/IsomorphicBatchWrite.java [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+ */
+public class IsomorphicBatchWrite implements LoadExecuteCallback {
+
+    private static final Logger LOG = LoggerFactory.getLogger(IsomorphicBatchWrite.class);
+
+    private static final String LABEL_PREFIX = "batch_write_";
+
+    private final long id;
+    private final TableId tableId;
+    private final String warehouseName;
```
[the remark this answers is not in the document]
Comment (banmoy) [reply fe/fe-core/src/main/java/com/starrocks/load/batchwrite/IsomorphicBatchWrite.java:57]: If warehouse is renamed, the client should also change the warehouse name, and a new `IsomorphicBatchWrite` should be created

File: fe/fe-core/src/main/java/com/starrocks/load/batchwrite/StatusOr.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import com.starrocks.thrift.TStatus;
+import com.starrocks.thrift.TStatusCode;
+
+/**
+ * Abstract class representing a status or a value.
+ * The value is only present if the status is OK.
+ *
+ * @param <T> The type of the value.
+ */
+public abstract class StatusOr<T> {
```
[the remark this answers is not in the document]
Comment (banmoy) [reply fe/fe-core/src/main/java/com/starrocks/load/batchwrite/StatusOr.java]: done

File: fe/fe-core/src/main/java/com/starrocks/load/batchwrite/LoadExecutor.java (hunk 1 above)
Comment (banmoy) [reply fe/fe-core/src/main/java/com/starrocks/load/batchwrite/LoadExecutor.java]: done

**Review by wyb — APPROVED** [review]

**Review by kevincai — APPROVED** [review]

**Pull request merged**